dnc/util.py

```diff
@@ -25,21 +25,48 @@
 def batch_invert_permutation(permutations):
   """Returns batched `tf.invert_permutation` for every row in `permutations`."""
   with tf.name_scope('batch_invert_permutation', values=[permutations]):
-    unpacked = tf.unstack(permutations)
-    inverses = [tf.invert_permutation(permutation) for permutation in unpacked]
-    return tf.stack(inverses)
+    perm = tf.cast(permutations, tf.float32)
+    dim = int(perm.get_shape()[-1])
+    size = tf.cast(tf.shape(perm)[0], tf.float32)
+    delta = tf.cast(tf.shape(perm)[-1], tf.float32)
+    rg = tf.range(0, size*delta, delta, dtype=tf.float32)
+    rg = tf.expand_dims(rg, 1)
+    rg = tf.tile(rg, [1, dim])
+    perm = tf.add(perm, rg)
+    flat = tf.reshape(perm, [-1])
+    perm = tf.invert_permutation(tf.cast(flat, tf.int32))
+    perm = tf.reshape(perm, [-1, dim])
+    return tf.subtract(perm, tf.cast(rg, tf.int32))
 
 
 def batch_gather(values, indices):
   """Returns batched `tf.gather` for every row in the input."""
   with tf.name_scope('batch_gather', values=[values, indices]):
-    unpacked = zip(tf.unstack(values), tf.unstack(indices))
-    result = [tf.gather(value, index) for value, index in unpacked]
-    return tf.stack(result)
+    idx = tf.expand_dims(indices, -1)
+    size = tf.shape(indices)[0]
+    rg = tf.range(size, dtype=tf.int32)
+    rg = tf.expand_dims(rg, -1)
+    rg = tf.tile(rg, [1, int(indices.get_shape()[-1])])
+    rg = tf.expand_dims(rg, -1)
+    gidx = tf.concat([rg, idx], -1)
+    return tf.gather_nd(values, gidx)
 
 
 def one_hot(length, index):
   """Return an nd array of given `length` filled with 0s and a 1 at `index`."""
   result = np.zeros(length)
   result[index] = 1
   return result
+
+def reduce_prod(x, axis, name=None):
+  '''
+  Uses tf.cumprod and tf.gather_nd as a workaround to the poor performance of calculating tf.reduce_prod's gradient
+  on CPU.
+  '''
+  with tf.variable_scope(name or "c_reduce_prod"):
+    cp=tf.cumprod(x, axis, reverse=True)
```

Comment:
Can you add spaces around your equals assignment?

size = tf.shape(cp)[0]
idx1 = ... etc.

You do not need to do this for keyword args however:

dtype=tf.float32 is fine!